vcd/resource_vcd_vapp.go

```diff
@@ -44,104 +42,23 @@ func resourceVcdVApp() *schema.Resource {
 				Description: "The name of VDC to use, optional if defined at provider level",
 			},
-			"template_name": {
-				Type:        schema.TypeString,
-				Optional:    true,
-				ForceNew:    true,
-				Description: "The name of the vApp Template to use",
-				Deprecated: implicitVmInVappDeprecation +
-					" Use vcd_vapp_vm.template_name instead",
-			},
-			"catalog_name": {
-				Type:        schema.TypeString,
-				Optional:    true,
-				Description: "The catalog name in which to find the given vApp Template",
-				Deprecated: implicitVmInVappDeprecation +
-					" Use vcd_vapp_vm.catalog_name instead",
-			},
-			"network_name": {
-				Type:        schema.TypeString,
-				Optional:    true,
-				ForceNew:    true,
-				Description: "Name of the network this vApp should join",
-				Deprecated: implicitVmInVappDeprecation +
-					" Use vcd_vapp_vm.network instead",
-			},
-			"memory": {
-				Type:        schema.TypeInt,
-				Optional:    true,
-				Description: "The amount of RAM (in MB) to allocate to the vApp",
-				Deprecated: implicitVmInVappDeprecation +
-					" Use vcd_vapp_vm.memory instead",
-			},
-			"cpus": {
-				Type:        schema.TypeInt,
-				Optional:    true,
-				Description: "The number of virtual CPUs to allocate to the vApp",
-				Deprecated: implicitVmInVappDeprecation +
-					" Use vcd_vapp_vm.cpus instead",
-			},
-			"ip": {
-				Type:        schema.TypeString,
-				Optional:    true,
-				Computed:    true,
-				Description: "The IP to assign to this vApp",
-				Deprecated: implicitVmInVappDeprecation +
-					" Use vcd_vapp_vm.network instead",
-			},
-			"storage_profile": {
-				Type:        schema.TypeString,
-				Optional:    true,
-				Description: "Storage profile to be used by the vApp",
-				Deprecated: implicitVmInVappDeprecation +
-					"use vcd_vapp_vm.storage_profile instead",
-			},
 			"description": {
 				Type:        schema.TypeString,
 				Optional:    true,
 				Description: "Optional description of the vApp",
 			},
-			"initscript": {
-				Type:        schema.TypeString,
-				Optional:    true,
-				ForceNew:    true,
-				Description: "A script to be run only on initial boot",
-				Deprecated: implicitVmInVappDeprecation +
-					" Use vcd_vapp_vm.initscript instead",
-			},
 			"metadata": {
 				Type:     schema.TypeMap,
 				Optional: true,
 				// For now underlying go-vcloud-director repo only supports
 				// a value of type String in this map.
 				Description: "Key value map of metadata to assign to this vApp. Key and value can be any string.",
 			},
-			"ovf": {
-				Type:        schema.TypeMap,
-				Optional:    true,
-				Description: "Key value map of ovf parameters to assign to VM product section",
-				Deprecated: implicitVmInVappDeprecation +
-					" Use guest_properties in this resource or vcd_vapp_vm.guest_properties instead",
-			},
 			"href": {
 				Type:        schema.TypeString,
 				Computed:    true,
 				Description: "vApp Hyper Reference",
 			},
-			"power_on": {
```

Comment:
Wou wou wou :) Why are we removing `power_on`?
1. It has not been deprecated.
2. This how 2.9 docs read:
> power_on - (Optional) A boolean value stating if this vApp should be powered on. Default is true
3. How can user control power on/off of vApp without this parameter?

Reply:
I reviewed `power_on` and brought back part of it's functionality. Though it works only in one case: you have created vApp and have VMs, then on update you can change value from `power_on=false` (change default value from true->false) to `power_on=true`. This due that vApp resource doesn't have VM's and power_on is meaningless - doesn't do anything. You can see in UI it doesn't allow to change state if vApp in `resolved` state and doesn't have VMs.

Reply:
I don't like this. This property has a specific functionality: tell the vApp to power on. When we do so, all the VMs will also be powered on. If we remove it, we are taking away an essential functionality. If it doesn't work correctly when the vApp is empty, we can adjust to that in the code.

Reply:
Well, as I mentioned currently this functionality works only in one case - on update if you already have VMs. On create it won't work cause you can't have have VMs until vApp is created. Another issue is that VMs have power_on value which will conflict with this value as they do similar things. Also additional issue here https://github.com/vmware/terraform-provider-vcd/issues/537. In summary I haven't changed `power_on` functionality - it is the same as it was, just made it `deprecated` that in future we will work on different approach how to handle actions.